### ISSUE TITLE
[UX] Add "tags" to the modules page and in .info files

### ISSUE THREAD

**Issue by jenlampton** [backdrop/backdrop-issues#37]
In order to group modules intelligently, let's debate whether we should use package, tags, or something else. 

Here's a compromise to move forward on tags discussed during today's UX meeting:

* Include tags[] in the .info file and display them on the modules page.
* Remove additional UI fields for "Tags" and "Module Source".
* Make it so that the search box includes the tags when doing a search.

And here are the follow-up issues:

* [ ] https://github.com/backdrop/backdrop-issues/issues/3146 Add checkboxes for module source: core/contrib/custom
* [ ] https://github.com/backdrop/backdrop-issues/issues/3147 Add a drop-down to browse by Tag
* [ ] https://github.com/backdrop/backdrop-issues/issues/3148 Alphabetize by module name and remove package groupings

Current PR screenshot: 
<img width="1251" alt="screen shot 2019-01-01 at 11 14 49 am" src="https://user-images.githubusercontent.com/397895/50575628-7cc4ec00-0db6-11e9-846d-f2371088521e.png">

---
~~PR by @docwilmot: https://github.com/backdrop/backdrop/pull/900~~
~~PR by @docwilmot: https://github.com/backdrop/backdrop/issues/1839~~
~~PR by @jenlampton: https://github.com/backdrop/backdrop/pull/2110~~
~~PR by @docwilmot: https://github.com/backdrop/backdrop/pull/2234~~
PR by @jenlampton: https://github.com/backdrop/backdrop/pull/2293

Pipe-dream Screenshot - shows filters for keyword / tag / core vs contrib
![screen shot 2018-04-29 at 10 26 49 pm](https://user-images.githubusercontent.com/397895/39415816-7ef400f0-4bfc-11e8-9bf1-210dfba0c361.png)

**Comment by quicksketch** [comment]
@docwilmot filed a PR for this at https://github.com/backdrop/backdrop/pull/900. Maybe we could get a screenshot?

**Comment by docwilmot** [comment]
Its a start!

![modules localhost 2](https://cloud.githubusercontent.com/assets/1042283/7695457/9d18a290-fdb5-11e4-9021-ba5e28406582.png)
![modules localhost 1](https://cloud.githubusercontent.com/assets/1042283/7695448/90b65042-fdb5-11e4-9cef-23cbf6d401fb.png)
![modules localhost](https://cloud.githubusercontent.com/assets/1042283/7695446/89765426-fdb5-11e4-923f-c39414893369.png)

**Comment by quicksketch** [comment]
Could we make it so users don't have to guess tag names by providing a drop-down list of checkboxes next to the search? Something similar to what we have here at GitHub, when you add/remove labels. That may be more work than we're interested in here.

I haven't weighed in on this issue at all, but the addition of the search box plus hiding dependencies has solved all the problems I've experienced on this page. The screenshot demonstrates a great example of *new* usability issues we're introducing. If you had installed the "Examples" project, you're either concerned with turning on Examples (for which you could use the search or tags) or you'd want to ignore them entirely. Skipping the entire group of examples is no longer possible (previously you'd scroll past the group or collapse it).

**Comment by docwilmot** [comment]
I agree. 

And that would be perfectly reasonable if all groups of modules came nicely packaged together like examples did, but when you have a module which could be media, could be user interface, maybe field-ish, it could be anywhere, so we end up with arbitrary groupings which you can never scroll past or collapse, because anything could be in there. So packaging them becomes useless.

And these days I never look through packages, because if you have a fast search (which could easily repackage them for you based on tags), arbitrary packaging becomes even more useless. 

Therefore if useless, why do it. Thats the point of this.

**Comment by sentaidigital** [comment]
What @docwilmot is describing is pretty close to a faceted search.  The Module Filter module used vertical tabs to filter the modules with a single-selection, single-value "faceted search"*.

99% of the time, I use the search field in Module Filter to find the module I'm trying to install.  However, from time to time, I will use the vertical tabs when I know I need to enable a couple modules that I already know are in a group.   Typically, LDAP related or Views related.

A faceted search would be more useful with more properties, which requires better metadata about the modules.  This is something I'd like to see added to Backdrop.  In particular, I'd like to see "suggests" and "supports" module tags in the info files.  

"Suggests" would note other modules that greatly enhance the usability of this one, but isn't strictly speaking required.  For example, Better Select converts standard select boxes to checkboxes, and would include "suggests[] = checkall" in the info file.

"Supports" would be for modules that provide enhancements to other modules.  For example, if a PayPal module included a hook_views_data() to make the PayPal transaction data available to Views, it would say "supports[] = views".  Again, Views is not required for this module to work.

These two tags may be more useful in Project Browser when a site builder is shopping for modules to install.

* is it really faceted if it is limited to one selection from one property?

**Comment by docwilmot** [comment]
> better metadata about the modules. This is something I'd like to see added to Backdrop.

@sentaidigital you're singing my song. See https://github.com/backdrop-ops/backdropcms.org/issues/100

**Comment by docwilmot** [comment]
As requested, a PR for testing the concept. For tags I borrowed from this (slightly modified) list on Drupal:
- Administration
- Commerce/Advertising
- Community
- Content
- Content Access Control
- Content Display
- Database Drivers
- Developer
- Drush
- E-commerce
- Education
- Evaluation/Rating
- Event
- Examples
- Features Package
- Fields
- File Management
- Filters and Editors
- Games and Amusements
- Import and Export
- JavaScript Utilities
- Language and Translation
- Location
- Mail
- Media
- Mobile
- Multilingual
- Multisite
- Novelty
- Organic Groups
- Paging
- Path Management
- Performance and Scalability
- Project management
- RDF
- Rules
- Search
- Security
- SEO
- Site Navigation
- Statistics
- Spam Prevention
- Syndication
- Taxonomy
- Theme Enhancements
- Third-party Integration
- User Access and Authentication
- User Management
- Utility

**Comment by docwilmot** [comment]
@klonos I modified your core/contrib toggle code and used it.

**Comment by docwilmot** [comment]
Also the toggle code is imperfect (understatement) since JQuery and I have never been close friends. 

But I think this PR holds up getting more info on BackdropCMS's modules listing because of the chicken and egg situation: we dont need code on BackdropCMS for categorizing modules, because there is no such  info in modules' `.info` files; BUT there's no point adding more info to module files because BackdropCMS.org's module list doesnt have any way to use this info!!

**Comment by Al-Rozhkov** [comment]
I like core/contrib switch and I like alphabetical order. Tags implementation kind of unusual for me for now. But I'm just used to packages, it is just a matter of habit.

Also great to see tags in Project Browser.

**Comment by Al-Rozhkov** [comment]
![modules-page](https://cloud.githubusercontent.com/assets/16973264/25066191/aff04854-2230-11e7-922d-c16aa5bb05e7.jpg)

We can improve a look by making filters inline like on Manage content page. What do you think?

**Comment by Al-Rozhkov** [comment]
![sortable-name-column](https://cloud.githubusercontent.com/assets/16973264/25066230/81648698-2231-11e7-89da-abdf15a50936.jpg)

Also "Name" column should by sortable to provide reverse order.

**Comment by Graham-72** [comment]
It is my understanding that this issue is about the order in which modules are listed on the modules page and to me this seems possibly a separate issue from providing some means of categorising the listing on backdropcms.org, where we have hundreds of modules to search through to find one to meet a particular need.
For the listing of modules installed on a site I would prefer to have a separation of the core and the contributed modules. I would like to be able to see at a glance which contributed modules I have installed, and I often refer to the available updates page to see this quickly.
When searching through available contrib modules I would find it useful to be able to list, for example, all modules that help provide photo-galleries, or all that provide a form of access control. I fear there must be a lengthy set of such possible classifications but something is needed to help us cope with the growing list of cotrib modules. Or perhaps we need an online 'ask klonos' service with encyclopedic knowledge of what is available 😃

**Comment by docwilmot** [comment]
@Graham-72 this issue is definitely related to categorization of modulus on b.org. The most logical way to have categorization there is to have these categories listed in info files. This PR redesigns the modules list but also introduces these tags for core modules. This is similar for example to when we introduced the `type = module ` string: we had to start with core modules first. 

The tags method is a better option than the existing "packages" method IMHO mainly because tags allow each module to be in several different "packages" at once. So this PR is actually an extension of current functionality rather than a change. 

The core vs contrib switch allows you to split those two groups easily, while still being able to search by tag and text even after separating core from contrib. 

Finally with this PR we can provide exactly this detailed categorisation you need on b.org. Once this data is also in contrib modules, you should be able to search as you need. The hope is that this will be step one to getting there. Almost as good as a virtual @klonos.

**Comment by Graham-72** [comment]
@docwilmot thanks for the added explanation. I was reading through all the above and just feeling that two different needs were getting muddled. I haven't yet had a proper look at your PR but I like the sound of it and the use of tags. It sounds as though you meet both needs. 👍

**Comment by klonos** [comment]
> ...Or perhaps we need an online 'ask klonos' service with encyclopedic knowledge of what is available 😃

😆 ...I spent too much time on the GitHub queues of Backdrop. I admit it. It is my pass time 😄 ...I'm away from my country with my brother. We *intentionally* do not own a TV (wasting time on youtube is enough as it is). We work most of the day and by the end of it, we have more work to do (fixing computers off-site for drop-off next day) and as most of you know our families are back in our country. This situation makes it stressful and depressing thoughts become overwhelming. Keeping our brains occupied with other things helps a great deal. For me these things are Backdrop 😄 ❤️

**Comment by klonos** [comment]
...I'm lost! Where's the PR? ...or is it just a mockup?

**Comment by Al-Rozhkov** [comment]
> ...I'm lost! Where's the PR? ...or is it just a mockup?

No :) PR is real: https://github.com/backdrop/backdrop/pull/1839

**Comment by klonos** [comment]
I guess https://github.com/backdrop/backdrop/pull/900 is all we have.

A few things:

- @Al-Rozhkov yes, inline filter controls 👍 
- Yes, I still think that a core/contrib toggle is very useful. I still think that this should not be a tag, but a separate thing of its own: #498
- Also, @Al-Rozhkov if your proposal about switching to button groups #2643 we should make this feature use that.
- @Graham-72 it is not just about the order things are shown on the modules list; it's also about being able to filter that list. It is also about being able to filter the list in the Project Browser and on b.org.
- The dropdown with the tags will be populated from tags in the .info files. In the modules list this will be scanning only the modules available on the current setup (the filesystem: things under `core/modules` and root `/modules`), while the list in b.org will be aggregated from what's in contrib from GitHub (plus whatever we have in core as @docwilmot said). The Project Browser can then pull that list from b.org.
- We will need to figure out how to standardize that list somehow to some extend: handle different spelling and capitalizations as well as synonyms etc.
- The list from d.org is a great start. Good on ya for pulling that @docwilmot 👍 
- For ease of use for those that prefer typing and that might already know what tag they're looking for, I suggest that the instant search filter also searches within tags.
- I do not like those big-@ss multiselect fields when they get inline with the rest of the single-line controls, but then again standard dropdowns won't allow selecting multiple options (tags in our case). So eventually, once we have #1800 in place, we should make that tags drop-down a better-select.

**Comment by klonos** [comment]
Thanx @Al-Rozhkov ...added the latest PR link to the issue summary 😉 ...on to the PR sandbox to try and break things 😄

**Comment by klonos** [comment]
> ...adding the latest PR link to the issue summary 😉

... looks like someone beat me to it :smile

**Comment by klonos** [comment]
> @klonos I modified your core/contrib toggle code and used it.

I've completely forgotten about that 😄 ...good one.

**Comment by klonos** [comment]
Not sure, but perhaps upon toggling the core/contrib switch, we should rebuild the tags dropdown menu to exclude categories with no results.

**Comment by klonos** [comment]
Small glitch: typing anything in the instant filter after a category is selected invalidates the core/contrib selection.

**Comment by klonos** [comment]
> - For ease of use for those that prefer typing and that might already know what tag they're looking for, I suggest that the instant search filter also searches within tags.

Steps to reproduce:

1. Select "Development" from the categories list -> Why is devel.module missing from the list (devel_generate is listed)?
2. Open another tab so that you can keep the list of modules from step 1 above as a reference.
3. Instead of selecting "Development" from the categories list, type "develop" -> most of the modules from the list in step 1 are missing (devel.module is there though 😕 ).

**Comment by klonos** [comment]
Great work @docwilmot 👍 ...thanks for bringing this back to life!

**Comment by klonos** [comment]
...and since this is not API-breaking from what I saw, can we please bring it back to 1.x-future?

**Comment by docwilmot** [comment]
Interesting fact: this is currently the oldest open issue in this queue. 
> can we please bring it back to 1.x-future?

Up to @jenlampton 

And re all the search and toggle issues, as mentioned I am not good at JQuery; usually what happens next is @quicksketch takes a look, has a good laugh, and fixes it. Hoping for that this time too.

If this is removed from the 2.x milestone, then I'll fix tests and other non-JS matters and get it as far as I can get it.
> Great work @docwilmot

Thanks, and thanks everyone

**Comment by klonos** [comment]
> And re all the search and toggle issues, as mentioned I am not good at JQuery; usually what happens next is @quicksketch takes a look, has a good laugh, and fixes it. Hoping for that this time too.

😄

**Comment by Graham-72** [comment]
With the ever-increasing number of contrib modules I think this should be scheduled ASAP and not wait until release 2.x-future.

**Comment by Dinornis** [comment]
👍 Thanks everyone, this is a great improvement.

**Comment by Dinornis** [comment]
Running the `Enable/disable core module and confirm table creation/deletion` tests on the PR site generated 57 fails and 3 exceptions.

**Comment by jenlampton** [comment]
Lets tag it 1.8 and see how much we can get done this cycle. Anything that we can't do now we can create a follow-up for and schedule that for 2.x.

**Comment by jenlampton** [comment]
We have 2 weeks until code freeze, and this is going to be a tricky issue. Bumping to 1.9 since there's still not a pull request ready for review and revision yet.

**Comment by jenlampton** [comment]
I'd like to tackle this issue in conjunction with https://github.com/backdrop-ops/backdropcms.org/issues/442 since they both go hand in hand. Since I'd like to implement something soon (in the 1.x cycle for Backdrop) I'd like to focus on adding `tags[]` to a module's info file, and importing that info to backdropcms.org to improve the search in both places. We can keep `project` as it is in 1.x, and if/when we want to remove the `project` groupings from 2.x we can move that direction, then.

**Comment by olafgrabienski** [comment]
@jenlampton I've had a look at the sandbox. My impression is that the `Keyword` filter works very good, and the `Module source` checkboxes work just fine.

The `Tag` filter works also quite good but I missed some results there. Here are a few examples:

- While the tag *Translation* returns "Content Translation", "Language" and "Locale", the tag *Language and Translation* returns only "Language" and "Locale", so "Content Translation" is missing here.
- The tag *Site Navigation* returns "Administration bar" and "Book", but "Menu" is missing.
- After having installed Honeypot, its tag *Spam control* was missing in the `Tag` filter list. (Typing *Spam* in the `Keyword` input worked perfectly though.)

**Comment by jenlampton** [comment]
> While the tag Translation returns "Content Translation", "Language" and "Locale", the tag Language and Translation returns only "Language" and "Locale", so "Content Translation" is missing here.

Sounds like some modules are missing some tags. An easy fix!

> The tag Site Navigation returns "Administration bar" and "Book", but "Menu" is missing.

Same thing, menu might be missing.

> After having installed Honeypot, its tag Spam control 

Are you sure honeypot has support for tags already? I don't see any tags on https://github.com/backdrop-contrib/honeypot/blob/1.x-1.x/honeypot.info. It looks like it has a package, maybe we should automatically make those tags as well?  I'll have a look at that...

**Comment by jenlampton** [comment]
> however I did find an issue that I can't actually enable or disable any of the modules

Ha! :) those should work, I bet I broke them.

**Comment by jenlampton** [comment]
@hosef I fixed the module saving problem, nice catch. I'm working on cleaning up the tags now.

**Comment by jenlampton** [comment]
Okay @laryn and @olafgrabienski I took a stab at reworking all the tags for all the core modules. I've documented everything here: https://api.backdropcms.org/documentation/module-packages-and-tags

I've also made it so that hidden and required modules to not add their tags to the select list on the modules page, so anything you see in that list should return results.

I think this is ready for another review - and since naming things is hard - feedback on the tags is also welcome :)

**Comment by olafgrabienski** [comment]
@jenlampton Wanted to review via the PR sandbox site but unfortunately I get a "HTTP ERROR 500" on its `admin/modules` page.

**Comment by olafgrabienski** [comment]
> I've documented everything here: https://api.backdropcms.org/documentation/module-packages-and-tags

One small thing: In the **Tags** list, "Account Management" is listed twice (at the very beginning and next to the last item).

**Comment by klonos** [comment]
@jenlampton, [3 years ago](https://github.com/backdrop/backdrop-issues/issues/37#issuecomment-67805262) ^^^:

> We don't have the time to add a new organization method like "tags" for 1.0.0. But we can certainly evaluate that for 2.x.

Seems we are doing it sooner than that! 👍 😉 

I have tested the PR on the sandbox and it looks really great! Good job @jenlampton 👍 ...we should be accounting for custom modules besides core/contrib though. I have filed a PR against your branch in your fork of Backdrop (I hope that this is the right way to do it): https://github.com/jenlampton/backdrop/pull/3

...as I am explaining in the PR summary, I wish that there was a better way to distinguish contrib vs custom modules, but for now I am assuming that any custom module will not have the `project`, `version`, `timestamp` bit of info in the "Added by Backdrop CMS packaging script on yyyy-mm-dd" block in their .info files.

What else I would like to see implemented:

- Group by package/List alphabetically radios. This would allow people to get rid of the fieldsets generated by the `package` property in the .info files. We can bike shade which option should be the default.
- ~~Add "Custom" checkbox~~ DONE! ...sorta 😄 ...see my PR
- Consider mirroring the general menu structure ("Structure" modules, "People" modules, "Content" modules). [Suggestion](https://github.com/backdrop/backdrop-issues/issues/37#issuecomment-24450166) by @cafuego 
- Enabled/Disabled/All filter
- Recently installed/enabled/updated modules (#532)
- Replace the "Tag" drop-down with a Select2 menu, once that is in core (#1800)
- Minor UX WTF: unselecting all core/contrib/custom checkboxes shows an empty table. I had some JS code at some time (PIE) that was ensuring that at least one checkbox was selected. Perhaps implement that in this checkbox combo. Either that, or @Al-Rozhkov's #2643 :)
- Consider adding a count next to the core/contrib/custom checkboxes that gets updated as the user types and things get filtered out(?).
- Another minor UX WTF (related to the point above): in a fresh installation (or in the odd case where one builds their site with Backdrop core only), the core/contrib/custom switch makes no sense and simply clutters the UI. Consider counting number of non-core modules, and if 0, then hide this filter entirely (or just whichever checkbox that has 0 count).
- Provide a reset buttons

@jenlampton: 

> ... It looks like it has a package, maybe we should automatically make those tags as well?

Yes please :)

@laryn:

> Agreed on making package into a tag if no tags are defined in a module.

...and yes, this sounds reasonable.

...we should get this in for 1.10! All these can be follow-ups. When I find some time, I will create a META (if one does not already exist), and create/link separate tickets.

Pheeeew! I started this review yesterday morning, and it is now 4am. Boy is life distracting :)

**Comment by jenlampton** [comment]
> Wanted to review via the PR sandbox site but unfortunately I get a "HTTP ERROR 500" on its admin/modules page.

Hm, me too. Works locally...  Maybe time for a new sandbox. closing issue...

**Comment by jenlampton** [comment]
Oh not *this* issue silly...

**Comment by jenlampton** [comment]
> Consider mirroring the general menu structure ("Structure" modules, "People" modules, "Content" modules). Suggestion by @cafuego

I tried to do this a little bit with the tag names. I wanted to use language people had already seen in the admin menu to help them filter the module list.

**Comment by klonos** [comment]
...I have left a comment in my PR, but bottom line is that yes, all these can be follow-ups :)

Need to go 😴 💤

**Comment by jenlampton** [comment]
Thanks @docwilmot and @Gormartsen for helping me locate what was going on with the sandbox!  The problem was that I had converted `#core` to `core` to prevent it from looking so much like a form API property, but since the value was a string instead of an array, Backdrop couldn't handle that change. 

@kolons -- you're going to love this -- since I had to change that property anyway, I took some code from your PR to make it into `#source` instead. Now it contains either `core` or `contrib` and I also added support for `custom`, only show the options for contrib or custom if there are contrib or custom modules present, and provide a count of each.

[The sandbox](https://github.com/backdrop/backdrop/pull/2110#issuecomment-385835446) is working again, for anyone interested in testing!

**Comment by olafgrabienski** [comment]
>> After having installed Honeypot, its tag Spam control

> Are you sure honeypot has support for tags already?

In the description text I had seen: "Tags: Spam control". Maybe some other information had been used as a tag. Anyway, above I reported that the tag `Spam control` was missing in the Tag filter list, and now it's there, so that has been fixed in the meantime.

**Comment by olafgrabienski** [comment]
I've found another issue: When I go to the *Modules* page, of the "Module source" checkboxes `Backdrop core` and `Contributed projects` only the core box is checked but the dropdown reflects also the contributed projects. If I choose a tag like `Spam control` (from *Honeypot*), I get no results. To get a result, I have to check the `Contributed projects` box first.

I'd suggest to check `Backdrop core` *and* `Contributed projects` by default.

**Comment by docwilmot** [comment]
I'm a bit disappointed by one thing this final version though. Why are we keeping the package fieldsets when we have tags now, instead of a simple alphabetical list? Are we really going to ask developers to add a package **and** tags? Thats not very logical. And also, if we dont require packages, will **all** contrib end up in a big "Other" package fieldset? And if we do accept packages wont we end up with the same disorganized mess of packages as D7 did? 
I propose that the search filter alone makes package fieldsets superfluous, and its even more so with  the core/contrib filter and the tags filter.
A simple alpha list of modules is what we should do.

**Comment by quicksketch** [comment]
UX-related improvements are not strictly limited from being merged after feature freeze (which was yesterday), so we could still get this in for 1.10 in the next two weeks.

But I'd like to collect feedback on the usefulness of this feature. I tried out the sandbox and I'm not entirely convinced. While it is good that modules can now be in multiple categories (tags), there are a huge number of categories and what is in them seems unpredictable.

> I'd suggest to check Backdrop core and Contributed projects by default.

This bug still seems to be present in the current version where only the "core" checkbox is checked on initial page load.

Tests are also failing as some tests seem unable to turn on their test modules.

I accidentally found that hitting the "Enter" key while in the search box submits the form and has very strange effects. For example I typed "System" and hit enter, and I got this:

![image](https://user-images.githubusercontent.com/100206/39543782-017955ce-4e01-11e8-98ff-ba85ff030bdc.png)

But the bizarre thing is that all of those modules should already be enabled. I don't know why Backdrop thinks it's enabling anything new. Prior to this patch, hitting enter while in the search box would not enable/disable any modules but would reload the page with a query string such as `?search=System`, and maintain the search value.

In a recent issue we made it so that the search box included module names, description, and their categories. But I found that the search does not include the new tags. For example, selecting the "Mail" tag shows "Contact" and "Email" modules. But searching for "Mail" only includes "Email" and not "Contact".

So this still needs some work yet.

**Comment by klonos** [comment]
@docwilmot yeah sorry, I am one of these people that still find the ability/feature of grouping things together (rather then an alpha list) to be very useful. I would not mind it if we made that a "Group by package" checkbox, and had it be unticked OOTB (which would be a blessing for people such as yourself) ...but so long as this option is remembered across sessions (so that people such as myself can tick it once, and have that stick).

@jenlampton yay for using `#source` instead of `#core` in code! 😄 ...and thanks for adding the counts in 👍

**Comment by docwilmot** [comment]
I wanted this for two reasons:
- while working on project Installer it became obvious that there was no way to allow searching for projects by category without adding more metadata into module files, and this was step one to get this done
- the package system is a mess; I think we're thinking of this in a nice new site where everything is neatly categorized, but in a mature site it breaks down horribly, and more so the idea that packages "organize" becomes laughable when the module creator thinks its a "Media enhancement" but you think its a "User interface". Tags, it seemed would allow both of you to be right.

But realistically, the main reason is the first. And for me what we need is a way forward for this tags-on-B.org initiative, with or without this PR.

**Comment by jenlampton** [comment]
Okay, I've updated the top issue with our compromize, and added links to the follow-up issues.

**Comment by jenlampton** [comment]
Un-assigning from myself so that if someone else wants to work on this one before I get back to it they can :)

**Comment by docwilmot** [comment]
Back again. Sadly. 
Hopefully this goes through this time.

**Comment by docwilmot** [comment]
Wait a minute! I just re-read the summary out the decision in the first post. I'm interpreting it belatedly as saying we only want the tags in and the other things as follow-up? I didn't bother waiting to follow-up though, since I didn't realise that was the plan. This latest PR does everything, fixes all the bugs, and tests are green.

**Comment by klonos** [comment]
@docwilmot yeah, we didn't have 100% consensus on all the suggestions proposed in this issue over time, that's why we have agreed to split this into 3 separate tasks. Main concerns are/were:

- we are adding too much "noise" in the exposed filters section
- the tags drop-down might get unwieldily over time (for example same/similar tags spelled differently), and having a drop-down with 100's of options is not useful nor user-friendly
- users need to know that a tag is available (which is not the case for Backdrop/Drupal newcomers) in order to be able to use it as a search filter.

Having said they above, I really like the result of your PR 👍 😄 ...

a) I was always on the fence re alpha listing vs grouping by package, so don't really mind one way or the other. I would prefer it if we had a "Group by package"/"List alphabetically" toggle in place though (which should remember the last state across sessions):

    ![screen shot 2018-07-02 at 5 53 44 am](https://user-images.githubusercontent.com/2423362/42138196-9a95e6e2-7dbc-11e8-9317-6b707f90afe6.png)

    ...this would decrease any "complains" about feature regression, and reduce "resistance" of implementing the alpha listing 😉

b) We should try to future-proof this as much as possible by "enforcing" some sort of casing in the tags, and then reducing duplicates. So for example "Site Navigation", "Site navigation", and "site navigation" should all be resulting in a single tag. Not sure how we would deal with cases like "Development" and "Dev" though. Hairy situation there, and I do not want to get further into it. I will personally be simply ignoring the drop-down if it gets too long, and thus annoying.

c) I really like the core/contrib/custom toggles, and so does @jenlampton. We find that it provides a quick overview/count of the modules under each one of these groups, and a quick way to toggle the list between them. I know that @quicksketch found this to be too much "noise", but if we switch to alpha listing, not having the toggle would be a serious feature regression IMO. Besides, we made sure that the toggles are rendered only if there are modules of the respective type found in the installation 😉 

I honestly would hate to keep delaying this issue any further, so if we still cannot reach consensus over which options should be available, can we just make these options in a new item under Configuration -> Administration (we are already using this path for the admin bar settings)?

**Comment by docwilmot** [comment]
Sorry I am causing us to revisit the bike shed after it has been decided. Honest mistake. But I feel strongly that having packages are a waste when we have the module filter. I'm fine with redoing the PR to remove the select list and the toggles, since the filter more or less substitutes for those. But I'm not in favor of putting packages back in.

**Comment by jenlampton** [comment]
@docwilmot the problem with removing packages -- especially if we don't have a select list for tags -- is that we have a UX regression for 2 use-cases:

1) the first-time user experience is degraded. There would be no organization of modules, and without experience or context, a massive list of things people don't already understand will be even less inviting.

2) the experience of a professional or experienced Drupal/Backdrop people when they are looking for an overview what modules are in-use on an existing site.  Want to know what feeds modules are in use? You're out of luck unless those modules contain a words you already know to search for. 

If we decide to remove the package groupings on this page, that can only be done in conjunction with adding the select list for tags. Removing them without providing anything that solves these 2 regressions is going to be making Backdrop into a less usable product.

**Comment by docwilmot** [comment]
@Graham-72 as I mentioned in https://github.com/backdrop/backdrop-issues/issues/1517#issuecomment-385013869, the overall aim is to get what you want done, but there are a lot of moving parts to get that done. We need to approach it as a combined project, otherwise it wont get far.

**Comment by Graham-72** [comment]
@docwilmot yes indeed. I was hoping to find a clear overall statement of requirements clearly written down somewhere, and trying to be clear about one important part of that.

**Comment by Graham-72** [comment]
@docwilmot I have just read https://github.com/backdrop-ops/backdropcms.org/issues/100 which is good. I particularly agree with your observation there (made in Dec 2016!)
>I think our project list is now large enough to need categorization. I just tried finding an ecommerce module; Paypal doesnt show up when you search "commerce", and Simple Cart doesnt show if you search "paypal".

When can we start formalising and using these new entries in .info files?

I would also like to see an entry for 'status'.

**Comment by jenlampton** [comment]
Assigning to myself. I'm about half-way through a reroll of the tags-only approach in the hopes we can get tags in 1.11, but I'm not sure I'm going to get it there today... But maybe by the 15th!  I've got what I have now up in https://github.com/backdrop/backdrop/pull/2293

**Comment by jenlampton** [comment]
Okay, I've pushed what I have to https://github.com/backdrop/backdrop/pull/2293. It contains:

- [x] Include tags[] in the .info file and display them on the modules page.
- [x] Remove additional UI fields for "Tags" and "Module Source".
- [x] Make it so that the search box includes the tags when doing a search.

And here are the follow-up issues:

  #3146 Add checkboxes for module source: core/contrib/custom
  #3147 Add a drop-down to browse by Tag
  #3148 Alphabetize by module name and remove package groupings

**Comment by klonos** [comment]
Thanks @jenlampton for pushing this forward 👍 ...I think that splitting this into smaller issues is the way to go. Also 👍 for documenting the tags being used in core modules 👍

I have checked the PR sandbox and found everything to be working as expected. I did not have any issues with core/contrib modules while using the filter as others have reported ... @herbdool care to elaborate on which modules and which terms you had trouble with so that I can then double-check/confirm?

I have left some comments about minor things here and there in the PR.

**Comment by klonos** [comment]
...by the way, I've noticed that you have left bits of the code that add the `core`/`contrib` classes. 👍 on that too. These will be useful even if we don't end up implementing #3146

**Comment by herbdool** [comment]
@klonos you mentioned me but I am not sure the context. Sure it was me?

**Comment by klonos** [comment]
@herbdool yep: https://github.com/backdrop/backdrop/pull/2293#issuecomment-418359528 😄

**Comment by herbdool** [comment]
@klonos I took another look and I can't recreate the issue I noticed.

**Comment by klonos** [comment]
@jenlampton I have pulled your changes and tested locally. When trying to enable the devel module (manually installed by copying and extracting under `/modules` - not via the Project Installer). I got this:

<img width="669" alt="screen shot 2018-11-08 at 11 37 44 pm" src="https://user-images.githubusercontent.com/2423362/48199214-98f86e00-e3af-11e8-9b40-2973df3c43f9.png">

...wondering if it has to do with https://github.com/backdrop/backdrop/pull/2293/files#r226666753

**Comment by jenlampton** [comment]
Sounds like you may not have a fresh install? All those projects should
already be enabled.

**Comment by klonos** [comment]
@jenlampton I have tested this twice already:

1. cloned your fork: https://github.com/jenlampton/backdrop
2. switched to the `37-module-tags` branch
3. started fresh installation
4. first thing to do is install devel module and then try to enable it from the UI

**Comment by klonos** [comment]
...I have tested the same, but this time tried installing devel via the Project Installer. There was no message about enabling additional (already enabled) modules. SO this is limited to installing "manually" by copying things into `/modules` dir. I hope that this helps.

**Comment by klonos** [comment]
Thanks for taking the time to post this status summary @docwilmot, it really helps 👍

...even if we don't end up having the time to implement all the prerequisites for this to happen in 1.12.0, I would still like to see an instant core/contrib/custom toggle added to the modules page.

**Comment by jenlampton** [comment]
> we need @quicksketch and @jenlampton to create the changes on Backdrop CMS and project module first ...

I think we may have a chicken and egg problem here. I've been holding off on any changes to Backdrop CMS or project module because there's no data to work with. This issue adds the data, which will unblock those issues. 

> ... the prerequisites for this to happen in 1.12.0 ...

There are no prerequisites for this to happen in 1.12.0. It all has to get done, and this seems like the simplest place to start :)

**Comment by docwilmot** [comment]
Well, actually, not really @jenlampton. Firstly, core modules dont get packaged as contrib releases do, so this PR wont help. Secondly, suppose you got into the Project module changes and needed some contrib modules to test it, we can easily create some dummy modules with tags for test purposes, or I can certainly add tags to some of the ports I've done; theres about 50 of those. So we have enough chickens.

I think the process needs to be: 
- get done the changes in Project module that detect tags (and maybe other `.info` file metadata at the same time, see https://github.com/backdrop-ops/backdropcms.org/issues/100)
- determine the mapping that we need for displaying fields with those tags on the main site
- update project browser to serve that data
- update Installer to use that data

**Comment by docwilmot** [comment]
Yes, agreed there is no reason not to finish this issue, only that I've realised it accomplishes little in the end. 
I'm hoping though that my point is clear, that to actually start categorizing our projects, the points I raise above make sense.

**Comment by jenlampton** [comment]
> that to actually start categorizing our projects, the points I raise above make sense.

Yes, fair enough. The good news though is that neither changing backdropcms.org or project module needs to be held up on backdrop release milestones :)

**Comment by docwilmot** [comment]
@jenlampton I've had think. How about you study https://github.com/backdrop-ops/backdropcms.org/issues/100 and actually create the fields on the project nodes to store this data. Once that happens you could leave those fields hidden from the front end until you find the time to actually create a proper front end UI. Meanwhile, the rest of us can continue looking into how to modify `borg_github` to fill those fields with data on project release events, and also to serve that data to Project Browser Server and Installer; since it would also be invisible to site installations until we update Installer in a new Backdrop core release, we can also test to our hearts content. 

I had a look at the code for project module. Unless I'm mistaken, to populate your fields we'd need to implement `hook_github_project_validate_release()` with code like:
```
  $project_node->field_project_tags[LANGUAGE_NONE][0] = array('value' => 'user experience');
  $project_node->field_project_tags[LANGUAGE_NONE][1] = array('value' => 'views');
  $project_node->field_project_documentation_url[LANGUAGE_NONE][0] = array('value' => 'github');
  $project_node->field_project_maintainers[LANGUAGE_NONE][0] = array('value' => 'jenlampton');
  etc
```
If @quicksketch can verify this, I'd happily PR after you @jenlampton have created those fields.

**Comment by jenlampton** [comment]
@klonos I've pushed a new PR that addresses most of the comments you left on the last one, and the problem you reported is also fixed :) Your comment about required modules actually steered me in the correct direction! Thank you!

**Comment by klonos** [comment]
Thanks @jenlampton I have tested the PR sandbox (although it could use a refresh), and I have also left a couple of comments in the PR. Other than than RTBC from me 👍

**Comment by jenlampton** [comment]
Comments addresses, new sandbox created, and ready for a final (?) review: https://github.com/backdrop/backdrop/pull/2293

**Comment by jenlampton** [comment]
Pushed an updated PR, tests should be passing now: https://github.com/backdrop/backdrop/pull/2293

**Comment by jenlampton** [comment]
Rebased, and pushed an updated PR.

**Comment by quicksketch** [comment]
PR looks good. Minor feedback: https://github.com/backdrop/backdrop/pull/2293#pullrequestreview-188586348

**Comment by quicksketch** [comment]
I've merged https://github.com/backdrop/backdrop/pull/2293 into 1.x for 1.11.0. Thanks everyone! It's a trimmed down implementation with _just_ tags and adding that into the filter search. This sets us up for having tags as part of Project Browser while also exposing this data into the Backdrop UI.
